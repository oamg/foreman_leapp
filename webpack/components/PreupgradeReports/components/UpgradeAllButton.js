import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';

import { idsForInvocation } from '../PreupgradeReportsHelpers';

const UpgradeAllButton = ({ preupgradeReports, postUrl, csrfToken }) => {
  const { hostIds } = idsForInvocation(preupgradeReports);

  return (
    <form action={postUrl} method="post">
      <Button type="submit">{__('Run Upgrade')}</Button>
      <input type="hidden" name="authenticity_token" value={csrfToken} />
      <input type="hidden" name="feature" value="leapp_upgrade" />
      {hostIds.map(hostId => (
        <input type="hidden" name="host_ids[]" key={hostId} value={hostId} />
      ))}
    </form>
  );
};

UpgradeAllButton.propTypes = {
  preupgradeReports: PropTypes.array.isRequired,
  postUrl: PropTypes.string.isRequired,
  csrfToken: PropTypes.string.isRequired,
};

export default UpgradeAllButton;
