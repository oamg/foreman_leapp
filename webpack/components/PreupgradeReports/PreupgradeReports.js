import React, { useState } from 'react';
import MessageBox from 'foremanReact/components/common/MessageBox';
import { LoadingState, Row } from 'patternfly-react';
import PropTypes from 'prop-types';

import PreupgradeReportsList from '../PreupgradeReportsList';
import UpgradeAllButton from './components/UpgradeAllButton';
import EntriesFilter from './components/EntriesFilter';
import FixSelectedButton from './components/FixSelectedButton';

import {
  flattenEntries,
  isEmpty,
  anyEntriesFixable,
  filterEntries,
  idsForInvocationFromEntries,
} from './PreupgradeReportsHelpers';

import NoReports from './components/NoReports';

const PreupgradeReports = ({
  preupgradeReports,
  csrfToken,
  newJobInvocationUrl,
}) => {
  const [filterType, setFilterType] = useState('title');
  const [filterValue, setFilterValue] = useState('');
  const [checked, setChecked] = useState([]);

  const onFilterValueChange = value => {
    setFilterValue(value);
  };

  const onFilterValueClear = () => setFilterValue('');

  const onFilterTypeChange = value => {
    onFilterValueClear();
    setFilterType(value);
  };

  const isSelected = entry => checked.some(item => item.id === entry.id);

  const anySelected = checked.length > 0;

  const toggleSelected = (entry, isEntrySelected) => {
    if (isEntrySelected) {
      setChecked(checked.filter(item => item.id !== entry.id));
    } else {
      setChecked([entry, ...checked]);
    }
  };

  return (
    <React.Fragment>
      <Row>
        <div className="col-md-8">
          <EntriesFilter
            filterType={filterType}
            onFilterTypeChange={onFilterTypeChange}
            filterValue={filterValue}
            onFilterValueChange={onFilterValueChange}
          />
        </div>
        <div className="col-md-4">
          <div className="btn-toolbar pull-right">
            <FixSelectedButton
              postUrl={newJobInvocationUrl}
              disabled={!anyEntriesFixable(preupgradeReports) || !anySelected}
              csrfToken={csrfToken}
              ids={idsForInvocationFromEntries(checked)}
            />
            <UpgradeAllButton
              postUrl={newJobInvocationUrl}
              csrfToken={csrfToken}
              preupgradeReports={preupgradeReports}
            />
          </div>
        </div>
      </Row>
      <PreupgradeReportsList
        allEntries={filterEntries(
          filterType,
          filterValue,
          flattenEntries(preupgradeReports)
        )}
        isSelected={isSelected}
        toggleSelected={toggleSelected}
      />
    </React.Fragment>
  );
};

const withLoadingState = Component => props => {
  const { error, loading, preupgradeReports, reportsExpected } = props;

  if (!isEmpty(error)) {
    return (
      <MessageBox
        key="preupgrade-reports-error"
        icontype="error-circle-o"
        msg={`Could not retrieve data: ${error.statusText} - ${error.errorMsg}`}
      />
    );
  }

  return (
    <LoadingState loading={loading}>
      {preupgradeReports.length > 0 ? (
        <Component {...props} />
      ) : (
        <NoReports reportsExpected={reportsExpected} />
      )}
    </LoadingState>
  );
};

PreupgradeReports.propTypes = {
  preupgradeReports: PropTypes.array.isRequired,
  csrfToken: PropTypes.string.isRequired,
  newJobInvocationUrl: PropTypes.string.isRequired,
};

export default withLoadingState(PreupgradeReports);
