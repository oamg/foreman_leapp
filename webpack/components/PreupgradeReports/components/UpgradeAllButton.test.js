import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import UpgradeAllButton from './UpgradeAllButton';

import { preupgradeReports } from '../__tests__/PreupgradeReports.fixtures';

const fixtures = {
  'should render': {
    preupgradeReports,
    postUrl: '/job_invocations/new',
    disabled: false,
    title: 'Button title',
    csrfToken: 'abcd',
  },
};

describe('UpgradeAllButton', () =>
  testComponentSnapshotsWithFixtures(UpgradeAllButton, fixtures));
