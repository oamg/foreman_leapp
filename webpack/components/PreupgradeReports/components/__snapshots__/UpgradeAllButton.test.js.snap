// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpgradeAllButton should render 1`] = `
<form
  action="/job_invocations/new"
  method="post"
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="default"
    disabled={false}
    type="submit"
  >
    Run Upgrade
  </Button>
  <input
    name="authenticity_token"
    type="hidden"
    value="abcd"
  />
  <input
    name="feature"
    type="hidden"
    value="leapp_upgrade"
  />
</form>
`;
